engine/src/main/java/org/terasology/entitySystem/event/AbstractValueModifiableEvent.java

```diff
@@ -44,31 +47,33 @@ public void add(float amount) {
         modifiers.add(amount);
     }
 
+    /**
+     * @deprecated please use {@link #postAdd(float)} instead.
+     */
+    @Deprecated
     public void addPostMultiply(float amount) {
+        postAdd(amount);
+    }
+
+    public void postAdd(float amount) {
         postModifiers.add(amount);
     }
 
+    /**
+     * Calculates the result value from the base value and given modifiers and multipliers.
+     * <p>
+     * The value is calculated based on the following formula:
+     * <pre>
+     * result = max(0, <baseValue> + Σ <modifier> * Π <multiplier> + Σ <postModifier>)
+     * </pre>
+     *
+     * <emph>The result value is guaranteed to be non-negative!</emph>
+     */
     public float getResultValue() {
-        // For now, add all modifiers and multiply by all multipliers. Negative modifiers cap to zero, but negative
-        // multipliers remain.
-
-        float result = baseValue;
-        TFloatIterator modifierIter = modifiers.iterator();
-        while (modifierIter.hasNext()) {
-            result += modifierIter.next();
-        }
-        result = Math.max(0, result);
-
-        TFloatIterator multiplierIter = multipliers.iterator();
-        while (multiplierIter.hasNext()) {
-            result *= multiplierIter.next();
-        }
-
-        final TFloatIterator postModifierIter = postModifiers.iterator();
-        while (postModifierIter.hasNext()) {
-            result += postModifierIter.next();
-        }
-        return result;
+        //TODO(skaldarnar): Based on a discussion in https://github.com/MovingBlocks/Terasology/pull/4063 we may want
+        // to lift the guarantee/restriction that the result value needs to be non-negative. Systems are still free to
+        // apply this restriction if needed.
+        return Math.max(0, baseValue + modifiers.sum() * product(multipliers) + postModifiers.sum());
```

Comment:
Thanks to @4Denthusiast for pointing this out in #4094

```suggestion
        return Math.max(0, (baseValue + modifiers.sum()) * product(multipliers) + postModifiers.sum());
```